File: AEDS II/TPs/TP3 (J)/ListaSequencial.java
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.Scanner;

public class ListaSequencial{
    public class Jogador{
        private int id;
        private String nome;
        private int altura;
        private int peso;
        private String universidade;
        private int anoNascimento;
        private String cidadeNascimento;
        private String estadoNascimento;
        static String[] allStrings;
        
        
        public Jogador(int nId, String nNome, int nAltura, int nPeso, String nUniversidade, int nAnoNascimento, String nCidadeNascimento, String nEstadoNascimento){
            this.id = nId;
            this.nome = nNome;
            this.altura = nAltura;
            this.peso = nPeso;
            this.universidade = nUniversidade;
            this.anoNascimento = nAnoNascimento;
            this.cidadeNascimento = nCidadeNascimento;
            this.estadoNascimento = nEstadoNascimento;
        }

        public Jogador(){
            this.id = -1;
            this.nome = "";
            this.altura = -1;
            this.peso = -1;
            this.universidade = "";
            this.anoNascimento = -1;
            this.cidadeNascimento = "";
            this.estadoNascimento = "";
        }

        public Jogador(int id){
            this();
            this.ler(id);
        }

        public String getNome(){
            return this.nome;
        }
        public void setNome(String nNome){
            this.nome = nNome;
        }

        public int getId(){
            return this.id;
        }
        public void setId(int nId){
            this.id = nId;
        }

        public int getAltura(){
            return this.altura;
        }
        public void setAltura(int nAltura){
            this.altura = nAltura;
        }

        public int getPeso(){
            return this.peso;
        }
        public void setPeso(int nPeso){
            this.peso = nPeso;
        }

        public String getUniversidade(){
            return this.universidade;
        }
        public void setUniversidade(String nUniversidade){
            this.universidade = nUniversidade;
        }

        public int getAnoNascimento(){
            return this.anoNascimento;
        }
        public void setAnoNascimento(int nAnoNascimento){
            this.anoNascimento = nAnoNascimento;
        }

        public String getCidadeNascimento(){
            return this.cidadeNascimento;
        }
        public void setCidadeNascimento(String nCidadeNascimento){
            this.cidadeNascimento = nCidadeNascimento;
        }

        public String getEstadoNascimento(){
            return this.estadoNascimento;
        }
        public void setEstadoNascimento(String nEstadoNascimento){
            this.estadoNascimento = nEstadoNascimento;
        }

        public Jogador clone(){
            Jogador res = new Jogador(this.id, this.nome, this.altura, this.peso, this.universidade, this.anoNascimento, this.cidadeNascimento, this.estadoNascimento);
            return res;
        }

        public void imprimir(){
            System.out.println("[" + this.id + " ## "+ this.nome + " ## " + this.altura + " ## " + this.peso + " ## " + this.anoNascimento + " ## " + this.universidade + " ## " + this.cidadeNascimento + " ## " + this.estadoNascimento + "]");
        }


        public static void getAllStrings(String nomeArq){
            try{
                RandomAccessFile raf = new RandomAccessFile(nomeArq, "r");
                String[] stringList = new String[10000];
                int i = 0;
                while(raf.getFilePointer() < raf.length()){
                    stringList[i] = raf.readLine();
                    i++;
                }
                allStrings = stringList;
                raf.close();
            }catch (IOException e){
                System.out.println("Erro ao abrir o arquivo");
            }
        }

        public void ler(int id){
            id += 1;
            if(allStrings == null){
                getAllStrings("/tmp/players.csv");
            }
            String[] propriedades = new String[8];

            String str = allStrings[id];
            int j = 0;
            propriedades[0] = "";
            for(int i=0;i<str.length();i++){
                if(str.charAt(i) == ','){
                    if(propriedades[j] == ""){
                        propriedades[j] = "nao informado";
                    }
                    j++;
                    propriedades[j] = "";
                }else{
                    propriedades[j] += str.charAt(i);
                }
            }

            if(propriedades[7] == ""){
                propriedades[7] = "nao informado";
            }

            this.id = Integer.parseInt(propriedades[0]);
            this.nome = propriedades[1];
            this.altura = Integer.parseInt(propriedades[2]);
            this.peso = Integer.parseInt(propriedades[3]);
            this.universidade = propriedades[4];
            this.anoNascimento = Integer.parseInt(propriedades[5]);
            this.cidadeNascimento = propriedades[6];
            this.estadoNascimento = propriedades[7];
        }
    }

    public class Lista{
        Jogador[] jogadores;
        int n;
        int total;

        public Lista(int total){
            this.jogadores = new Jogador[total];
            this.n = 0;
            this.total = total;
        }

        public void inserirInicio(Jogador jogador){
            if(n>=total){
                System.out.println("Erro");
                return;
            }
            this.n++;

            for(int i=n-1;i>=1;i--){
                this.jogadores[i] = this.jogadores[i-1];
            }
            this.jogadores[0] = jogador;
        }

        public void inserirFim(Jogador jogador){
            if(n>=total){
                System.out.println("Erro");
                return;
            }
            jogadores[n] = jogador;
            n++;
        }

        public void inserir(Jogador jogador, int pos){
            if(pos>n || pos<0 || n>=total){
                System.out.println("Erro");
                return;
            }
            for(int i=n-1;i>=pos;i--){
                jogadores[i+1] = jogadores[i];
            }
            jogadores[pos] = jogador;
            n++;
        }

        public Jogador removerInicio(){
            if(n==0){
                System.out.println("Erro");
                return null;
            }

            Jogador res = jogadores[0];

            for(int i=0;i<n-1;i++){
                jogadores[i] = jogadores[i+1];
            }
            n--;
            return res;
        }

        public Jogador removerFim(){
            if(n==0){
                System.out.println("Erro");
                return null;
            }
            n--;
            return jogadores[n];
        }

        public Jogador remover(int pos){
            if(pos>=n || pos<0 || n==0){
                System.out.println("Erro");
                return null;
            }

            Jogador res = jogadores[pos];
            for(int i=pos;i<n-1;i++){
                jogadores[i] = jogadores[i+1];
            }

            n--;
            return res;
        }

        public void print(){
            for(int i=0;i<n;i++){
                System.out.print("[" + i + "] ## " + jogadores[i].getNome() + " ## " + jogadores[i].getAltura() + " ## " + jogadores[i].getPeso() + " ## " + jogadores[i].getAnoNascimento() + " ## " + jogadores[i].getUniversidade() + " ## " + jogadores[i].getCidadeNascimento() + " ## " + jogadores[i].getEstadoNascimento() + " ##\n");
            }
        }

        public void command(String str){
            if(str.contains("II")){
                String[] sides = str.split(" ");
                int id = Integer.parseInt(sides[1]);
                Jogador j = new Jogador(id);
                inserirInicio(j);
            }else if(str.contains("IF")){
                String[] sides = str.split(" ");
                int id = Integer.parseInt(sides[1]);
                Jogador j = new Jogador(id);
                inserirFim(j);
            }else if(str.contains("I*")){
                String[] sides = str.split(" ");
                int pos = Integer.parseInt(sides[1]);
                int id = Integer.parseInt(sides[2]);
                Jogador j = new Jogador(id);
                inserir(j, pos);
            }else if(str.contains("RI")){
                Jogador j = removerInicio();
                System.out.println("(R) " + j.getNome());
            }else if(str.contains("RF")){
                Jogador j = removerFim();
                System.out.println("(R) " + j.getNome());
            }else if(str.contains("R*")){
                String[] sides = str.split(" ");
                int pos = Integer.parseInt(sides[1]);
                Jogador j = remover(pos);
                System.out.println("(R) " + j.getNome());
            }else{
                int id = Integer.parseInt(str);
                Jogador j = new Jogador(id);
                inserirFim(j);
            }
        }
    }

    
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        ListaSequencial ls = new ListaSequencial();
        Lista lista = ls.new Lista(10000);
        String in = sc.nextLine();
        while(!in.equals("FIM")){
            lista.command(in);
            in = sc.nextLine();
        }
        int qtdCom = Integer.parseInt(sc.nextLine());
        for(int i=0;i<qtdCom;i++){
            String str = sc.nextLine();
            lista.command(str);
        }

        lista.print();

        sc.close();
    }
}

File: AEDS II/TPs/TP3 (J)/listaSequencial.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Jogador{
    int id;
    char nome[50];
    int altura;
    int peso;
    char universidade[100];
    int anoNascimento;
    char cidadeNascimento[50];
    char estadoNascimento[50];
} Jogador;

void readFile(FILE* mignon, Jogador* jogadores){
    char linha[300];
    fscanf(mignon, "%s", linha);
    char props[8][100];
    int i = 0;
    fgets(linha, 300, mignon);
    while(!feof(mignon)){
        fgets(linha, 300, mignon);

        int j = 0;
        int a = 0;
        for(int k=0;k<strlen(linha);k++){
            if(linha[k] == ',' || linha[k] == '\n'){
                props[j][a] = '\0';
                j++;
                a = 0;
            }else{
                props[j][a] = linha[k];
                a++;
            }
        }

        for(int b=0;b<8;b++){
            if(strcmp(props[b], "") == 0){
                strcpy(props[b], "nao informado");
            }
        }

        jogadores[i].id = atoi(props[0]);
        strcpy(jogadores[i].nome, props[1]);
        jogadores[i].altura = atoi(props[2]);
        jogadores[i].peso = atoi(props[3]);
        strcpy(jogadores[i].universidade, props[4]);
        jogadores[i].anoNascimento = atoi(props[5]);
        strcpy(jogadores[i].cidadeNascimento, props[6]);
        strcpy(jogadores[i].estadoNascimento, props[7]);

        i++;
    }
    fclose(mignon);
}


void imprimir(Jogador jogador){
    printf("[%d ## %s ## %d ## %d ## %d ## %s ## %s ## %s]", jogador.id, jogador.nome, jogador.altura, jogador.peso, jogador.anoNascimento, jogador.universidade, jogador.cidadeNascimento, jogador.estadoNascimento);
}

typedef struct Lista{
    Jogador* arr;
    int n;
    int total;
} Lista;

Lista* newLista(){
    Lista* lista = (Lista*) malloc(sizeof(Lista));
    lista->arr = (Jogador*) malloc(10000 * sizeof(Jogador));
    lista->n = 0;
    lista->total = 10000;
    return lista;
}

Lista* inserirInicio(Lista* lista, Jogador jogador){
    if(lista->n == lista->total){
        printf("Erro ao inserir no inicio");
        exit(1);
    }
    for(int i=lista->n;i>0;i--){
        lista->arr[i] = lista->arr[i-1];
    }
    lista->arr[0] = jogador;
    lista->n++;
    return lista;
}

Lista* inserirFim(Lista* lista, Jogador jogador){
    if(lista->n == lista->total){
        printf("Erro ao inserir no fim");
        exit(1);
    }
    lista->arr[lista->n] = jogador;
    lista->n++;
    return lista;
}

Lista* inserir(Lista* lista, Jogador jogador, int pos){
    if(pos > lista->n || pos < 0 || lista->n == lista->total){
        printf("Erro ao inserir");
        exit(1);
    }

    for(int i=lista->n;i>pos;i--){
        lista->arr[i] = lista->arr[i-1];
    }

    lista->arr[pos] = jogador;
    lista->n++;
    return lista;
}

Lista* removerInicio(Lista* lista){
    if(lista->n == 0){
        printf("Erro ao remover no inicio");
        exit(1);
    }
    for(int i=0;i<lista->n-1;i++){
        lista->arr[i] = lista->arr[i+1];
    }
    lista->n--;
    return lista;
}

Lista* removerFim(Lista* lista){
    if(lista->n == 0){
        printf("Erro ao remover no fim");
        exit(1);
    }
    lista->n--;
    return lista;
}

Lista* remover(Lista* lista, int pos){
    if(pos >= lista->n || pos < 0 || lista->n == 0){
        printf("Erro ao remover");
        exit(1);
    }

    for(int i=pos;i<lista->n-1;i++){
        lista->arr[i] = lista->arr[i+1];
    }

    lista->n--;
    return lista;
}

void print(Lista* lista){
    for(int i=0;i<lista->n;i++){
        printf("[%d] ## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n", i, lista->arr[i].nome, lista->arr[i].altura, lista->arr[i].peso, lista->arr[i].anoNascimento, lista->arr[i].universidade, lista->arr[i].cidadeNascimento, lista->arr[i].estadoNascimento);
    }
}

void scanCommand(Lista* lista, Jogador* jogadores){
    char* command = (char*) malloc(3 * sizeof(char));
    scanf("%s", command);
    if(strcmp(command, "II") == 0){
        int id;
        scanf("%d", &id);
        inserirInicio(lista, jogadores[id]);
    }else if(strcmp(command, "IF") == 0){
        int id;
        scanf("%d", &id);
        inserirFim(lista, jogadores[id]);
    }else if(strcmp(command, "I*") == 0){
        int id, pos;
        scanf("%d", &pos);
        scanf("%d", &id);
        inserir(lista, jogadores[id], pos);
    }else if(strcmp(command, "RI") == 0){
        printf("(R) %s\n", lista->arr[0].nome);
        removerInicio(lista);
    }else if(strcmp(command, "RF") == 0){
        printf("(R) %s\n", lista->arr[lista->n-1].nome);
        removerFim(lista);
    }else if(strcmp(command, "R*") == 0){
        int pos;
        scanf("%d", &pos);
        printf("(R) %s\n", lista->arr[pos].nome);
        remover(lista, pos);
    }
    free(command);
}



int main(void){
    FILE* mignon = fopen("/tmp/players.csv", "r");
    Jogador* jogadores = (Jogador*) malloc(10000 * sizeof(Jogador));

    readFile(mignon, jogadores);

    Lista* lista = newLista();

    char* line = (char*) malloc(23 * sizeof(char));
    
    scanf("%s", line);

    while(strcmp(line, "FIM") != 0){
        inserirFim(lista, jogadores[atoi(line)]);
        scanf("%s", line);
    }

    int n;
    scanf("%d", &n);

    for(int i=0;i<n;i++){
        scanCommand(lista, jogadores);
    }

    print(lista);

    return 0;
}